File: typescript/packages/core/src/commands/builtin/dev/io.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { RAMAccessor } from '../../../accessor/ram.ts'
import type { IndexCacheStore } from '../../../cache/index/store.ts'
import { read, readRange, stat, stream } from '../../../core/dev/index.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandIO } from '../generic_bind/index.ts'
import { RAM_IO } from '../ram/io.ts'

async function* finiteStream(
  accessor: RAMAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  const data = await read(accessor, path, index)
  if (data.byteLength > 0) yield data
}

export const DEV_STREAMING_IO: CommandIO<RAMAccessor> = {
  ...RAM_IO,
  readBytes: read,
  readRange,
  readStream: stream,
  stat,
}

export const DEV_IO: CommandIO<RAMAccessor> = {
  ...DEV_STREAMING_IO,
  readStream: finiteStream,
}

File: python/mirage/commands/builtin/dev/io.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import dataclasses
from collections.abc import AsyncIterator

from mirage.accessor.ram import RAMAccessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.commands.builtin.ram.io import IO as _RAM_IO
from mirage.core.dev.read import read as _read
from mirage.core.dev.stat import stat as _stat
from mirage.core.dev.stream import read_stream as _read_stream
from mirage.types import PathSpec


async def _finite_read_stream(
        accessor: RAMAccessor,
        path: PathSpec,
        index: IndexCacheStore = NULL_INDEX) -> AsyncIterator[bytes]:
    data = await _read(accessor, path, index)
    if data:
        yield data

# /dev is a RAM mount whose read and stat know the two synthetic
# character devices; every other slot is RAM's. Commands that consume a
# whole input use the refusing read operation, while the two bounded
# streaming commands opt into the endless source below.
STREAMING_IO = dataclasses.replace(
    _RAM_IO,
    stat=_stat,
    read_bytes=_read,
    read_range=_read,
    read_stream=_read_stream,
)
IO = dataclasses.replace(STREAMING_IO, read_stream=_finite_read_stream)

resolve_glob = IO.resolve_glob
